eZ/Bundle/EzPublishCoreBundle/FieldType/RichText/Converter/Html5Input.php

```diff
@@ -22,4 +23,27 @@ public function __construct($stylesheet, ConfigResolverInterface $configResolver
         $customStylesheets = $customStylesheets ?: array();
         parent::__construct($stylesheet, $customStylesheets);
     }
+
+    /**
+     * Method checks if document contains any links with '[' or ']' sign an replaces it to its html entities version.
+     *
+     * @param DOMDocument $document
+     *
+     * @return DOMDocument
+     */
+    public function convert(DOMDocument $document)
+    {
+        // EZP-27562 - links containing [ or ] sign results in xml validation error
+        foreach ($document->getElementsByTagName('a') as $link) {
+            $href = $link->getAttribute('href');
+            if ($href !== '') {
+                $href = str_replace('[', '%5B', $href);
+                $href = str_replace(']', '%5D', $href);
```

Comment:
What if there are other forbidden characters in the href? I'd rather try to use `urlencode`. The problem with it is it's designed to work with url parts, not the whole url, so some weird looking parsing would be required:

```php
$parsed = parse_url($href);
if (!empty($parsed['path'])) {
    $encodedItems = [];
    foreach (explode('/', $parsed['path']) as $pathItem) {
        $encodedItems[] = urlencode($pathItem);
    }
    $parsed['path'] = implode('/', $encodedItems);
}
```
The other problem with this is that the opposite to `parse_url` `http_build_url` does not exist in core PHP, it's a pecl extension...
I still think covering just this case is not enough. However, maybe we can do it simpler?
Any POV @andrerom @Nattfarinn?

Reply:
I will take care of this issue.
 
>  What if there are other forbidden characters in the href?

@alongosz Here is BNF grammar: https://stackoverflow.com/a/41895124 (copied from original RFCs).  

> I assume there are some form of validation further down, so felt wrong to validate it and escape to workaround bugs in RelaxNg

If we would like to fix this issue without hacks, we must change the definition of href attribute in one of the two ways: 

1. Change data type from `anyURI` to `string` in definition of `href` attribute (It's already allowed to put value like `<script>alert('BOOM!')</script>`...)
2. Use a custom regular expression to validate attribute value

Reply:
We can't allow XSS :)
So either go with 2. or accept that RFC is the Bible and encode all URI parts prior saving. The latter one is troublesome/unclean because it requires splitting path (as path `/` separator should remain unencoded), which can lead to spaghetti code.

Really not sure which way to go...

Reply:
> We can't allow XSS :)

@alongosz `href` value is properly escaped when internal XML is transformed to HTML, so there is no risk of XSS here. 

@andrerom WDYT ?

Reply:
> internal XML is transformed to HTML,

output you are talking about here? Well we prefer to disallow it already on input to be on the safe side. Or did you mean something else?